File: Tests/ClipboardCoreTests/ClipStoreTests.swift
import Foundation
import Testing
@testable import ClipboardCore

@Suite("ClipStore")
struct ClipStoreTests {
    @Test("stores text clips newest first and reloads them from disk")
    func storesTextClipsNewestFirstAndReloads() throws {
        let clock = TestClock(Date(timeIntervalSince1970: 1_000))
        let store = try ClipStore(storageDirectory: temporaryDirectory(), now: clock.now)

        let first = try store.addText("first")
        clock.advance(by: 10)
        let second = try store.addText("second")

        #expect(store.items.map(\.id) == [second.id, first.id])

        let reloaded = try ClipStore(storageDirectory: store.storageDirectory, now: clock.now)
        #expect(reloaded.items.map(\.text) == ["second", "first"])
    }

    @Test("deduplicates text content by moving an existing clip to the top")
    func deduplicatesTextClips() throws {
        let clock = TestClock(Date(timeIntervalSince1970: 2_000))
        let store = try ClipStore(storageDirectory: temporaryDirectory(), now: clock.now)

        let first = try store.addText("repeat")
        clock.advance(by: 20)
        let duplicate = try store.addText("repeat")

        #expect(first.id == duplicate.id)
        #expect(store.items.count == 1)
        #expect(store.items.first?.createdAt == clock.now())
    }

    @Test("removes only unpinned clips older than seven days")
    func removesOnlyUnpinnedExpiredClips() throws {
        let start = Date(timeIntervalSince1970: 3_000)
        let clock = TestClock(start)
        let store = try ClipStore(storageDirectory: temporaryDirectory(), now: clock.now)

        let expired = try store.addText("expired")
        let pinned = try store.addText("pinned")
        try store.setPinned(pinned.id, true)
        clock.advance(by: (8 * 24 * 60 * 60) + 1)
        let fresh = try store.addText("fresh")

        try store.cleanupExpiredClips()

        #expect(store.items.map(\.id).contains(expired.id) == false)
        #expect(store.items.map(\.id).contains(pinned.id))
        #expect(store.items.map(\.id).contains(fresh.id))
    }

    @Test("stores image clips as files and reloads metadata")
    func storesImageClipsAsFiles() throws {
        let clock = TestClock(Date(timeIntervalSince1970: 4_000))
        let store = try ClipStore(storageDirectory: temporaryDirectory(), now: clock.now)
        let imageData = Data([0x89, 0x50, 0x4E, 0x47])

        let clip = try store.addImageData(imageData, fileExtension: "png", previewText: "Screenshot")

        #expect(clip.type == .image)
        #expect(clip.previewText == "Screenshot")
        #expect(try Data(contentsOf: clip.resolvedImageURL(baseDirectory: store.storageDirectory)) == imageData)

        let reloaded = try ClipStore(storageDirectory: store.storageDirectory, now: clock.now)
        #expect(reloaded.items.first?.type == .image)
        #expect(reloaded.items.first?.imagePath == clip.imagePath)
    }

    @Test("filters pinned clips across text and image items")
    func filtersPinnedClips() throws {
        let clock = TestClock(Date(timeIntervalSince1970: 5_000))
        let store = try ClipStore(storageDirectory: temporaryDirectory(), now: clock.now)

        let pinnedText = try store.addText("important text")
        let unpinnedText = try store.addText("normal text")
        let pinnedImage = try store.addImageData(Data([0x01, 0x02, 0x03]), fileExtension: "png", previewText: "Pinned image")

        try store.setPinned(pinnedText.id, true)
        try store.setPinned(pinnedImage.id, true)

        let pinnedClips = store.clips(matching: "", filter: .pinned)

        #expect(pinnedClips.map(\.id).contains(pinnedText.id))
        #expect(pinnedClips.map(\.id).contains(pinnedImage.id))
        #expect(pinnedClips.map(\.id).contains(unpinnedText.id) == false)
    }
}

private final class TestClock {
    private var current: Date

    init(_ current: Date) {
        self.current = current
    }

    func now() -> Date {
        current
    }

    func advance(by seconds: TimeInterval) {
        current = current.addingTimeInterval(seconds)
    }
}

func temporaryDirectory() throws -> URL {
    let url = FileManager.default.temporaryDirectory
        .appendingPathComponent("ClipboardSSTests-\(UUID().uuidString)", isDirectory: true)
    try FileManager.default.createDirectory(at: url, withIntermediateDirectories: true)
    return url
}

File: ClipboardSS.Windows/tests/ClipboardSS.Core.Tests/ClipStoreTests.cs
using ClipboardSS.Core.Models;
using ClipboardSS.Core.Storage;

namespace ClipboardSS.Core.Tests;

public sealed class ClipStoreTests
{
    [Fact]
    public void StoresNewestFirstAndReloadsFromDisk()
    {
        using var directory = new TemporaryDirectory();
        var clock = new TestClock(DateTimeOffset.FromUnixTimeSeconds(1_000));
        var store = new ClipStore(directory.Path, clock.Now);

        var first = store.AddText("first");
        clock.Advance(TimeSpan.FromSeconds(10));
        var second = store.AddText("second");

        Assert.Equal([second.Id, first.Id], store.Items.Select(item => item.Id));
        var reloaded = new ClipStore(directory.Path, clock.Now);
        Assert.Equal(["second", "first"], reloaded.Items.Select(item => item.Text));
    }

    [Fact]
    public void DeduplicatesByHashRefreshesTimestampAndPreservesIdentityAndPin()
    {
        using var directory = new TemporaryDirectory();
        var clock = new TestClock(DateTimeOffset.FromUnixTimeSeconds(2_000));
        var store = new ClipStore(directory.Path, clock.Now);
        var original = store.AddText("repeat");
        store.SetPinned(original.Id, true);
        clock.Advance(TimeSpan.FromSeconds(20));

        var duplicate = store.AddText("repeat");

        Assert.Equal(original.Id, duplicate.Id);
        Assert.Single(store.Items);
        Assert.True(store.Items[0].IsPinned);
        Assert.Equal(clock.Now(), store.Items[0].CreatedAt);
    }

    [Fact]
    public void MarkCopiedSetsTimestampAndMovesClipToFront()
    {
        using var directory = new TemporaryDirectory();
        var clock = new TestClock(DateTimeOffset.FromUnixTimeSeconds(3_000));
        var store = new ClipStore(directory.Path, clock.Now);
        var first = store.AddText("first");
        var second = store.AddText("second");
        clock.Advance(TimeSpan.FromMinutes(1));

        store.MarkCopied(first.Id);

        Assert.Equal([first.Id, second.Id], store.Items.Select(item => item.Id));
        Assert.Equal(clock.Now(), store.Items[0].LastCopiedAt);
    }

    [Fact]
    public void RemovesOnlyUnpinnedClipsOlderThanSevenDays()
    {
        using var directory = new TemporaryDirectory();
        var clock = new TestClock(DateTimeOffset.FromUnixTimeSeconds(4_000));
        var store = new ClipStore(directory.Path, clock.Now);
        var expired = store.AddText("expired");
        var pinned = store.AddText("pinned");
        store.SetPinned(pinned.Id, true);
        clock.Advance(TimeSpan.FromDays(8) + TimeSpan.FromSeconds(1));
        var fresh = store.AddText("fresh");

        store.CleanupExpiredClips();

        Assert.DoesNotContain(store.Items, item => item.Id == expired.Id);
        Assert.Contains(store.Items, item => item.Id == pinned.Id);
        Assert.Contains(store.Items, item => item.Id == fresh.Id);
    }

    [Fact]
    public void StoresImagesAsFilesDeduplicatesAndDeletesBackingFile()
    {
        using var directory = new TemporaryDirectory();
        var clock = new TestClock(DateTimeOffset.FromUnixTimeSeconds(5_000));
        var store = new ClipStore(directory.Path, clock.Now);
        byte[] imageData = [0x89, 0x50, 0x4e, 0x47];
        var clip = store.AddImageData(imageData, "png", "Screenshot");
        var imagePath = clip.ResolveImagePath(directory.Path);

        Assert.Equal(imageData, File.ReadAllBytes(imagePath));
        Assert.Equal(clip.Id, store.AddImageData(imageData, "jpg").Id);
        Assert.Single(Directory.GetFiles(System.IO.Path.Combine(directory.Path, "Images")));

        store.Delete(clip.Id);
        Assert.False(File.Exists(imagePath));
    }

    [Fact]
    public void PreviewSearchAndFiltersMatchSwiftBehavior()
    {
        using var directory = new TemporaryDirectory();
        var store = new ClipStore(directory.Path);
        var empty = store.AddText(" \n\t ");
        var searchable = store.AddText("First line\nSecret second line");
        var image = store.AddImageData([1, 2, 3], "png", "Receipt image");
        store.SetPinned(image.Id, true);

        Assert.Equal("Empty text", empty.PreviewText);
        Assert.Equal("First line Secret second line", searchable.PreviewText);
        Assert.Equal([searchable.Id], store.Clips("second", ClipFilter.Text).Select(item => item.Id));
        Assert.Equal([image.Id], store.Clips("receipt", ClipFilter.Image).Select(item => item.Id));
        Assert.Equal([image.Id], store.Clips(string.Empty, ClipFilter.Pinned).Select(item => item.Id));
    }

    private sealed class TestClock(DateTimeOffset current)
    {
        private DateTimeOffset _current = current;
        public DateTimeOffset Now() => _current;
        public void Advance(TimeSpan interval) => _current += interval;
    }
}
